File: 1250-longest-common-subsequence/longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # Tabulation method
        # either move t1 and check
        # or move t2 and check
        # or move both forward

        dp = [[0]*(len(text1)+1) for _ in range(len(text2)+1)]
        print(dp)


        for i in range(len(text2)):
            for j in range(len(text1)):
                dp[i+1][j+1] = ( 
                    (dp[i][j]+1) if text1[j] == text2[i]    #when the chars present in said indices are the same for the 2 strings, inc LCS by 1
                     else max(dp[i][j+1] , dp[i+1][j] )     #else max depicts best traversal of the 2 choices, increment index of i or j by 1 
                )

        return dp[-1][-1]

File: 1250-longest-common-subsequence/longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<short>> dp(text2.size()+1, vector<short>(text1.size()+1, 0));

        for( short i = 0; i<text2.size();i++){
            for (short j = 0; j<text1.size(); j++){
                // if the chars match
                if (text2[i] == text1[j]){
                    dp[i+1][j+1] = dp[i][j]+1;
                }
                else{
                     dp[i+1][j+1] = max(dp[i+1][j] , dp[i][j+1]);
                }
            }
        }
        return dp[text2.size()][text1.size()];
    }
};